File: 프로그래머스/0/120866. 안전지대/안전지대.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> board) {
    int answer = 0;
    int check_board[1001][1001]; // 외각은 사용하지 않을 것
    for (int i = 0; i < board.size(); i++)
    {
        for (int j = 0; j < board[0].size(); j++)
        {
            if (board[i][j] == 1)
            {
                check_board[i-1+1][j-1+1] = 1; // 외각을 사용하지 않기 때문에 각 인덱스 값의 +1 위치 사용
                check_board[i-1+1][j+1] = 1;
                check_board[i-1+1][j+1+1] = 1;
                check_board[i+1][j-1+1] = 1;
                check_board[i+1][j+1] = 1;
                check_board[i+1][j+1+1] = 1;
                check_board[i+1+1][j-1+1] = 1;
                check_board[i+1+1][j+1] = 1;
                check_board[i+1+1][j+1+1] = 1;
            }
            cout << board[i][j] << " ";
        }
        cout << endl;

    }
    cout << endl;
    for (int i = 0; i < board.size(); i++)
    {
        for (int j = 0; j < board[0].size(); j++)
        {
            if (check_board[i + 1][j + 1] != 1)
                answer++;
            cout << check_board[i + 1][j + 1] << " ";
        }
        cout << endl;
    }
    return answer;
}

File: 프로그래머스/0/120866. 안전지대/안전지대.py
def solution(board):
    answer = 0
    rows,cols=len(board),len(board[0])
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
    
    danger_map=[[0]*cols for _ in range(rows)]
    
    for i in range(rows):
        for j in range(cols):
            if board[i][j]==1:
                danger_map[i][j]=1
                for dx,dy in directions:
                    ni,nj=i+dx,j+dy
                    if 0<=ni<rows and 0<=nj<cols:
                        danger_map[ni][nj]=1
                
    for i in range(rows):
        for j in range(cols):
            if danger_map[i][j]==0:
                answer+=1
    
    return answer